diff --git a/MineSearch/MineSearch.Game/BaseCell.cs b/MineSearch/MineSearch.Game/BaseCell.cs
index 0f18fdb..e02cbf5 100644
--- a/MineSearch/MineSearch.Game/BaseCell.cs
+++ b/MineSearch/MineSearch.Game/BaseCell.cs
@@ -2,7 +2,7 @@
 
 namespace MineSearch.Game
 {
-    public abstract class BaseCell : ICell
+    public abstract class BaseCell : ModelBase, ICell
     {
         #region Properties
 
@@ -14,12 +14,34 @@ namespace MineSearch.Game
         /// <summary>
         /// Whether or not the cell has been flagged.
         /// </summary>
-        public bool Flagged { get; set; }
+        public bool Flagged
+        {
+            get { return _flagged; }
+            set
+            {
+                if (value != _flagged)
+                {
+                    _flagged = value;
+                    OnPropertyChanged();
+                }
+            }
+        }
 
         /// <summary>
         /// Whether or not the cell has been revealed.
         /// </summary>
-        public bool Revealed { get; set; }
+        public bool Revealed
+        {
+            get { return _revealed; }
+            set
+            {
+                if (value != _revealed)
+                {
+                    _revealed = value;
+                    OnPropertyChanged();
+                }
+            }
+        }
 
         #endregion
 
@@ -31,6 +53,8 @@ namespace MineSearch.Game
         #region Fields
 
         private readonly Point _coordinates;
+        private bool _flagged;
+        private bool _revealed;
 
         #endregion
     }
diff --git a/MineSearch/MineSearch.Game/MineSearch.Game.csproj b/MineSearch/MineSearch.Game/MineSearch.Game.csproj
index fe5d7a5..b6da1f5 100644
--- a/MineSearch/MineSearch.Game/MineSearch.Game.csproj
+++ b/MineSearch/MineSearch.Game/MineSearch.Game.csproj
@@ -48,6 +48,7 @@
     <Compile Include="MineCell.cs" />
     <Compile Include="MineSearchCellsFactory.cs" />
     <Compile Include="MineSearchGame.cs" />
+    <Compile Include="ModelBase.cs" />
     <Compile Include="Properties\AssemblyInfo.cs" />
     <Compile Include="SafeCell.cs" />
   </ItemGroup>
diff --git a/MineSearch/MineSearch.Wpf.Test/MineSearchGameViewModelTest.cs b/MineSearch/MineSearch.Wpf.Test/MineSearchGameViewModelTest.cs
index 20efac2..daf78db 100644
--- a/MineSearch/MineSearch.Wpf.Test/MineSearchGameViewModelTest.cs
+++ b/MineSearch/MineSearch.Wpf.Test/MineSearchGameViewModelTest.cs
@@ -22,13 +22,14 @@ namespace MineSearch.Wpf.Test
             var cells = _gameViewModel.Game.Cells;
             var viewModels = _gameViewModel.CellViewModels;
 
-            Assert.AreEqual(cells.Rows, viewModels.Rows);
-            Assert.AreEqual(cells.Columns, viewModels.Columns);
-
-            foreach (var cell in _gameViewModel.Game.Cells)
+            foreach (var row in viewModels)
             {
-                var viewModel = viewModels[cell.Coordinates.X, cell.Coordinates.Y];
-                Assert.AreEqual(cell, viewModel.Cell);
+                foreach (var viewModel in row)
+                {
+                    var coords = viewModel.Cell.Coordinates;
+                    var cell = cells[coords.X, coords.Y];
+                    Assert.AreEqual(cell, viewModel.Cell);
+                }
             }
         }
 
diff --git a/MineSearch/MineSearch.Wpf/MineSearch.Wpf.csproj b/MineSearch/MineSearch.Wpf/MineSearch.Wpf.csproj
index 8a6bbc7..a5c67dc 100644
--- a/MineSearch/MineSearch.Wpf/MineSearch.Wpf.csproj
+++ b/MineSearch/MineSearch.Wpf/MineSearch.Wpf.csproj
@@ -101,6 +101,10 @@
     <Compile Include="ViewModels\IMineSearchGameViewModel.cs" />
     <Compile Include="ViewModels\MainWindowViewModel.cs" />
     <Compile Include="ViewModels\MineSearchGameViewModel.cs" />
+    <Page Include="Views\UserControls\CellView.xaml">
+      <SubType>Designer</SubType>
+      <Generator>MSBuild:Compile</Generator>
+    </Page>
     <Page Include="Views\UserControls\MineSearchGameView.xaml">
       <SubType>Designer</SubType>
       <Generator>MSBuild:Compile</Generator>
@@ -113,6 +117,9 @@
       <DependentUpon>App.xaml</DependentUpon>
       <SubType>Code</SubType>
     </Compile>
+    <Compile Include="Views\UserControls\CellView.xaml.cs">
+      <DependentUpon>CellView.xaml</DependentUpon>
+    </Compile>
     <Compile Include="Views\UserControls\MineSearchGameView.xaml.cs">
       <DependentUpon>MineSearchGameView.xaml</DependentUpon>
     </Compile>
diff --git a/MineSearch/MineSearch.Wpf/ViewModels/IMineSearchGameViewModel.cs b/MineSearch/MineSearch.Wpf/ViewModels/IMineSearchGameViewModel.cs
index 615a022..1ce6e65 100644
--- a/MineSearch/MineSearch.Wpf/ViewModels/IMineSearchGameViewModel.cs
+++ b/MineSearch/MineSearch.Wpf/ViewModels/IMineSearchGameViewModel.cs
@@ -1,4 +1,5 @@
-﻿using System.Windows.Input;
+﻿using System.Collections.Generic;
+using System.Windows.Input;
 using MineSearch.Common;
 using MineSearch.Game;
 
@@ -24,6 +25,6 @@ namespace MineSearch.Wpf.ViewModels
         /// <summary>
         /// Matrix of cell view models.
         /// </summary>
-        IMatrix<ICellViewModel> CellViewModels { get; }
+        List<List<ICellViewModel>> CellViewModels { get; }
     }
 }
diff --git a/MineSearch/MineSearch.Wpf/ViewModels/MineSearchGameViewModel.cs b/MineSearch/MineSearch.Wpf/ViewModels/MineSearchGameViewModel.cs
index a697d20..6ed26a4 100644
--- a/MineSearch/MineSearch.Wpf/ViewModels/MineSearchGameViewModel.cs
+++ b/MineSearch/MineSearch.Wpf/ViewModels/MineSearchGameViewModel.cs
@@ -1,6 +1,6 @@
-﻿using System.Windows.Input;
+﻿using System.Collections.Generic;
+using System.Windows.Input;
 using Microsoft.Practices.Prism.Commands;
-using MineSearch.Common;
 using MineSearch.Common.ViewModels;
 using MineSearch.Game;
 
@@ -38,7 +38,7 @@ namespace MineSearch.Wpf.ViewModels
         /// <summary>
         /// Matrix of cell view models.
         /// </summary>
-        public IMatrix<ICellViewModel> CellViewModels
+        public List<List<ICellViewModel>> CellViewModels
         {
             get { return _cellViewModels; }
             private set
@@ -73,11 +73,15 @@ namespace MineSearch.Wpf.ViewModels
         {
             GameSettings = gameSettings;
             Game = new MineSearchGame(GameSettings);
-            var cellViewModels = new Matrix<ICellViewModel>(Game.Rows, Game.Columns);
-            foreach (var cell in Game.Cells)
+            var cellViewModels = new List<List<ICellViewModel>>(GameSettings.Rows);
+            for (int row = 0; row < GameSettings.Rows; row++)
             {
-                var viewModel = new CellViewModel(Game, cell);
-                cellViewModels[cell.Coordinates.X, cell.Coordinates.Y] = viewModel;
+                cellViewModels.Add(new List<ICellViewModel>(GameSettings.Columns));
+                for (int col = 0; col < GameSettings.Columns; col++)
+                {
+                    var cell = Game.Cells[row, col];
+                    cellViewModels[row].Add(new CellViewModel(Game, cell));
+                }
             }
             CellViewModels = cellViewModels;
         }
@@ -85,7 +89,7 @@ namespace MineSearch.Wpf.ViewModels
         #region Fields
 
         private IMineSearchGame _game;
-        private IMatrix<ICellViewModel> _cellViewModels;
+        private List<List<ICellViewModel>> _cellViewModels;
 
         #endregion
     }
diff --git a/MineSearch/MineSearch.Wpf/Views/UserControls/MineSearchGameView.xaml b/MineSearch/MineSearch.Wpf/Views/UserControls/MineSearchGameView.xaml
index b4b1103..8005e06 100644
--- a/MineSearch/MineSearch.Wpf/Views/UserControls/MineSearchGameView.xaml
+++ b/MineSearch/MineSearch.Wpf/Views/UserControls/MineSearchGameView.xaml
@@ -2,10 +2,30 @@
              xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
              xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
              xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" 
-             xmlns:d="http://schemas.microsoft.com/expression/blend/2008" 
+             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+             xmlns:userControls="clr-namespace:MineSearch.Wpf.Views.UserControls"
              mc:Ignorable="d" 
              d:DesignHeight="300" d:DesignWidth="300">
     <Grid>
-            
+        <Grid.Resources>
+            <DataTemplate x:Key="CellTemplate">
+                <userControls:CellView />
+            </DataTemplate>
+            <DataTemplate x:Key="RowDataTemplate">
+                <ItemsControl 
+                    ItemsSource="{Binding}"
+                    ItemTemplate="{DynamicResource CellTemplate}">
+                    <ItemsControl.ItemsPanel>
+                        <ItemsPanelTemplate>
+                            <StackPanel Orientation="Horizontal" />
+                        </ItemsPanelTemplate>
+                    </ItemsControl.ItemsPanel>
+                </ItemsControl>
+            </DataTemplate>
+        </Grid.Resources>
+        <ItemsControl
+            HorizontalAlignment="Center"
+            ItemsSource="{Binding CellViewModels}"
+            ItemTemplate="{DynamicResource RowDataTemplate}" />
     </Grid>
 </UserControl>
diff --git a/MineSearch/MineSearch.Wpf/Views/Windows/MainWindow.xaml.cs b/MineSearch/MineSearch.Wpf/Views/Windows/MainWindow.xaml.cs
index 259329d..afa8687 100644
--- a/MineSearch/MineSearch.Wpf/Views/Windows/MainWindow.xaml.cs
+++ b/MineSearch/MineSearch.Wpf/Views/Windows/MainWindow.xaml.cs
@@ -18,6 +18,7 @@ namespace MineSearch.Wpf.Views.Windows
             InitializeComponent();
 
             ViewModel = viewModel;
+            DataContext = ViewModel;
         }
     }
 }
